streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -92,69 +95,80 @@ public void validatePartitionExists(final TopicPartition topicPartition, final S
 
     @Override
     public void register(final StateRestorer restorer) {
-        if (restorer.offsetLimit() > 0) {
-            stateRestorers.put(restorer.partition(), restorer);
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+
+    public Collection<TopicPartition> restore() {
+        if (!initialized) {
+            initialize();
+        }
+
+        if (needsRestoring.isEmpty()) {
```

Comment:
For the former case, we should be transiting to `RUNNING` after the last `restore()` call so it will not be called again with `needRestoring` already empty, right?